### PR TITLE
Hangman Game is Developed and tested

### LINKED ISSUE #1297: [Add a Play]: Hangman Game
### What's the Play about?

I would like to develop Hangman game.

**Below is the description of the game:**

- Ask a question related to a word and let the user guess it.
- Limit users to a certain number of wrong guesses.
- Display parts of a hangman figure for each incorrect guess.
- If the hangman figure is fully formed, the user loses.

**Could you please assign me this issue?**

### What ReactJS concept will be used to create this play?

I will use React useState, useRef, useEffect hooks. 
It will  use react-dom and other libraries of the react.

### Record

- [X] I have checked the existing [issues](https://github.com/reactplay/react-play/issues)
- [X] I have read the [Contributing Guidelines](https://github.com/reactplay/react-play/blob/main/CONTRIBUTING.md)
- [X] I want to work on this issue

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ANKITy102**
> First thing, PLEASE READ THIS: [ReactPlay Code Review Checklist](https://github.com/reactplay/react-play/wiki/ReactPlay-Code-Review-Checklist)

**Description**
I have developed and thoroughly tested the Hangman game, and it's fully functional and ready for review.

**Game Features:**
- **Hint Display:** Users are presented with a hint for the word they need to guess.
- **Word Guessing:** Users are required to guess the word based on the given hint.
- **Incorrect Guess Handling:** Incorrect guesses lead to the gradual creation of the hangman figure.
- **Loss Condition:** If the hangman figure is completed, the user loses the game.
- **Win Condition:** Upon successful completion, the user can reset the game and receive a new word to guess.

**I have used styed components to build the components efficiently and responsive. So styled-components library is installed. No other libraries are installed.**

Fixes #1297 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

**Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.**

- All the prettier checks are passed. I used standard to avoid prettier errors.
- Code is also tested by the linter. 
- I have run the application on the React-Play site and it is working fine.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

# Screenshots or example output
Screenshot 2023-10-07 001903
Screenshot 2023-10-07 001924
Screenshot 2023-10-07 001936
Screenshot 2023-10-07 001959

### CODE REVIEW

**Comment by ANKITy102** [comment]
@joshi-kaushal 

Could you please review my changes and pr.

File: src/plays/hangman-game/components/Keyboard.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
+import { useEffect } from 'react';
+import React from 'react';
+import styled from 'styled-components';
+
+const KEYS = [
```
Comment (sudipkundu999) [inline src/plays/hangman-game/components/Keyboard.tsx:5]: move to constants

File: src/plays/hangman-game/components/Keyboard.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+const KeyContainer = styled.div`
+  align-self: stretch;
+`;
+
+const KeyboardGrid = styled.div`
+  display: grid;
+  grid-template-columns: repeat(auto-fit, minmax(50px, 1fr));
+  gap: 0.5rem;
+  width: 100%;
+`;
```
Comment (sudipkundu999) [inline src/plays/hangman-game/components/Keyboard.tsx:100]: Move all styled-components to `styles.css` and import & use them.

Benefits 
- the JS logic and styles will be separate. So, easier to debug specific problems.
- the respective lines on code on each files will be less

File: src/plays/hangman-game/components/Main.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
+import React, { useEffect, useState } from 'react';
+import styled from 'styled-components';
+import Drawing from './Drawing';
+import words from './wordList.json';
+import Keyboard from './Keyboard';
+import Word from './Word';
+
+const tryAgainIcon = (
```
Comment (sudipkundu999) [inline src/plays/hangman-game/components/Main.tsx:8]: move to constants

File: src/plays/hangman-game/components/wordList.json [hunk 4]
```diff
@@ -0,0 +1,51 @@
+[
```
Comment (sudipkundu999) [inline src/plays/hangman-game/components/wordList.json:1]: Good to not keep this file in `/components/` as it is not a react component.

Preferably use a `/constants/` and keep all data there.

File: src/plays/hangman-game/styles.css [hunk 5]
```diff
@@ -0,0 +1 @@
+/* enter stlyes here */
```
Comment (sudipkundu999) [inline src/plays/hangman-game/styles.css:1]: keep all styled-components here & export them to use in respective files

**Review by sudipkundu999 — CHANGES_REQUESTED** [review]

File: src/plays/hangman-game/components/Keyboard.tsx (hunk 2 above)
Comment (ANKITy102) [reply src/plays/hangman-game/components/Keyboard.tsx:100]: @sudipkundu999  done!

File: src/plays/hangman-game/components/wordList.json (hunk 4 above)
Comment (ANKITy102) [reply src/plays/hangman-game/components/wordList.json:1]: @sudipkundu999  done!

File: src/plays/hangman-game/styles.css (hunk 5 above)
Comment (ANKITy102) [reply src/plays/hangman-game/styles.css:1]: @sudipkundu999  done!

**Comment by ANKITy102** [comment]
@sudipkundu999 

I've implemented the requested changes and am now eagerly awaiting your feedback. 
Would you kindly review my revisions?

File: src/plays/hangman-game/components/Main.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        )}
+        {!(isWinner || isLoser) && (
+          <p
+            style={{
+              marginBottom: '-20px',
+              fontSize: '20px'
+            }}
+          >
+            <span style={{ fontWeight: 'bold' }}>Question: {hintToGuess} </span>
+          </p>
```
Comment (sudipkundu999) [inline src/plays/hangman-game/components/Main.tsx:65]: let's avoid inline styling and use styled components for consistency.

File: src/plays/hangman-game/constants/constants.tsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  't',
+  'u',
+  'v',
+  'w',
+  'x',
+  'y',
+  'z'
+];
+
+export const tryAgainIcon = (
```
Comment (sudipkundu999) [inline src/plays/hangman-game/constants/constants.tsx:32]: name could be changed to `TRY_AGAIN_ICON`

File: src/plays/hangman-game/components/Main.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    restartGame();
+    // eslint-disable-next-line
+  }, []);
+
+  return (
+    <BigContainer>
+      <Container>
+        <Title>Hangman</Title>
+
+        {!(isWinner || isLoser) ? (
```
Comment (sudipkundu999) [inline src/plays/hangman-game/components/Main.tsx:49]: - 4 places this variable is being used in the same file.
   extract it out as a common variable `const isWinnerOrLoser = isWinner || isLoser`
- or maybe a better name `isGameCompleted` or anything you like
- then you code should have a simple condition
`isGameCompleted ? < winner loser message, try again,... /> : < game drawing,... />`

**Review by sudipkundu999 — CHANGES_REQUESTED** [review]

File: src/plays/hangman-game/constants/constants.tsx (hunk 7 above)
Comment (ANKITy102) [reply src/plays/hangman-game/constants/constants.tsx]: @sudipkundu999  done!

**Comment by ANKITy102** [comment]
@sudipkundu999 

Thank you for your valuable feedback and guidance. I have incorporated your suggestions and pushed the necessary commits. Your continued input is greatly appreciated.

**Comment by sudipkundu999** [comment]
LGTM!
@priyankarpal can you please deploy?

**Comment by priyankarpal** [comment]
> LGTM!
> @priyankarpal can you please deploy?

I don't have access of vercel

**Comment by sudipkundu999** [comment]
> > LGTM!
> > @priyankarpal can you please deploy?
> 
> I don't have access of vercel

Netlify?

**Comment by priyankarpal** [comment]
> > > LGTM!
> > > @priyankarpal can you please deploy?
> > 
> > 
> > I don't have access of vercel
> 
> Netlify?

no

**Comment by ANKITy102** [comment]
@joshi-kaushal 
Could you please review and consider merging my pull request? Thank you!

**Comment by ANKITy102** [comment]
@priyankarpal 
Could you please review and consider merging my pull request? Thank you!

**Comment by priyankarpal** [comment]
> @priyankarpal Could you please review and consider merging my pull request? Thank you!

we will do it soon.

**Review by priyankarpal — DISMISSED** [review]

**Comment by atapas** [comment]
@ANKITy102 On it.

**Review by atapas — APPROVED** [review]
Looks Good to me.

**Review by sudipkundu999 — CHANGES_REQUESTED** [review]
Please, fix these styling issues
- The try again button does not look like it
    
    image


- The hangman figure has some styling issues
    
    image

**Comment by atapas** [comment]
@ANKITy102 any update?

**Comment by ANKITy102** [comment]
> @ANKITy102 any update?

@atapas I am having some stuff to complete. But I will fix the styling issue tonight without any delay. Sorry for the delay.

**Comment by ANKITy102** [comment]
@atapas 
I have fixed the styling issue. Below is the image for the same. 
<hr/>

### Fixed hangman figure styling issue

image

### Fixed restart icon style

image

**Review by atapas — APPROVED** [review]
Styles changes are looking good.

@sudipkundu999 please check.

**Review by sudipkundu999 — APPROVED** [review]
Looks Good 🚀

**Review by priyankarpal — APPROVED** [review]

**Pull request merged**